### PR TITLE
Rk/websockets comment updates

### PR DESCRIPTION
**Pull request by unicornzero**
This PR provides instant UI updates for comments:

* Implements websockets with Pusher
* Uses an auth endpoint to connect to private websocket channels
* Receives websocket updates for the creation of comments and archiving of comment threads, and immediately updates the ember data store to trigger updates in the UI

## Reviewing this PR

_**Is Covered by Automated Tests?: YES**_

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -133,5 +134,9 @@
     "hooks": {
       "pre-commit": "lint-staged"
     }
+  },
```
Comment (unicornzero) [inline package.json]: This seems the wrong place. They should be up in devDependencies, yes?

**Comment by samselikoff** [comment]
This is a little hacky, but you can use Mirage to generate the JSON:API payload.

At the end of your `createWebhookPayload` hook I tried this:

```js
let lastComment = server.schema.comments.find(lastExistingCommentId)
server.serializerOrRegistry.serialize(lastComment, {
  queryParams: { include: 'author,comment-thread' }
})
```

I think this produces the payload you want. I can explain what's going on:

- `lastComment` is a Mirage Model instance, so if you update the `createWebhookPayload` to actually create a new Mirage model, you could use that instead
- `server.serializerOrRegistry` is kinda private but kinda public API, and if it ever changes it would be mentioned in the changelog.
- The `serialize` method takes in a Model, and the second arg is a request, so we're passing in a request-like object with `queryParams` key simulating a request that's like `GET /comments/1?include=author,comment-thread`

I can chat more about this later but this does seem to produce the payload you're after. If you omit the request-like 2nd argument Mirage will use the default serializers that are set up and you'll get the serialized comment with ALL its relationships (which might be ok for your use case too).

File: mirage/factories/comment-thread.js [hunk 2]
```diff
@@ -5,7 +5,10 @@ import faker from 'faker';
 
 export default Factory.extend({
   type: REVIEW_COMMENT_TYPE,
-  createdAt: () => moment().subtract(22, 'hours'),
+  createdAt: () =>
+    moment()
+      .subtract(22, 'hours')
+      .format(),
```
Comment (unicornzero) [inline mirage/factories/comment-thread.js]: This stringifies the value, allowing it to work in my tests without me having to run JSON.stringify and JSON.parse on the objects.

File: mirage/factories/comment.js [hunk 3]
```diff
@@ -6,5 +6,5 @@ export default Factory.extend({
   body: () => faker.hacker.phrase(),
   author: association(),
   createdAt: () => moment().subtract(22, 'hours'),
-  commentThread: association('commentThread'),
+  commentThread: association(),
```
Comment (unicornzero) [inline mirage/factories/comment.js]: When testing something else (now deleted), I discovered the incorrect syntax here.

File: app/routes/organization.js [hunk 4]
```diff
@@ -41,6 +43,12 @@ export default Route.extend({
     }
   },
 
+  async _setupWebsocket(organization) {
+    if (await isUserMemberPromise(organization)) {
```
Comment (cadeParade) [inline app/routes/organization.js]: Don't need to await this anymore...rebase?

File: app/services/websocket.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+import utils from 'percy-web/lib/utils';
+import Pusher from 'pusher-js';
+
+export default Service.extend({
+  store: service(),
+
+  subscribeToOrganization(organization) {
+    const channelName = `private-organization-${organization.id}`;
+
+    if (this._client.channels[channelName]) {
```
Comment (cadeParade) [inline app/services/websocket.js]: Is it important that the key have any certain value, or only that the key be present in the `channels` hash?

File: app/services/websocket.js [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        cluster: config.APP.PUSHER_APP_CLUSTER,
+        authEndpoint: utils.buildApiUrl('websocketsAuth'),
+        auth: {
+          headers: {'X-CSRF-Token': csrfToken},
+        },
+      });
+    },
+    set(key, value) {
+      if (config.environment === 'test') {
+        return value;
```
Comment (cadeParade) [inline app/services/websocket.js]: Let's put that assert back in...we don't want anyone setting this value unless they're in a test.  https://api.emberjs.com/ember/release/functions/@ember%2Fdebug/assert

File: app/services/websocket.js (hunk 5 above)
Comment (unicornzero) [reply app/services/websocket.js]: My intention is to not subscribe to the same organization's channel multiple times.

File: app/services/websocket.js [hunk 7]
```diff
@@ -4,28 +4,45 @@ import config from '../config/environment';
...
-    });
+  disconnect() {
+    this._socket.disconnect();
   },
 
-  subscribe(channelName, event, callback) {
-    const channel = this._client.subscribe(channelName);
+  _bindEvents(channel, event, callback) {
     channel.bind(event, callback);
+    channel.bind('error', function(error) {
```
Comment (cadeParade) [inline app/services/websocket.js]: What does this do?

Comment (unicornzero) [reply app/services/websocket.js]: If Pusher sends an error, such as if we run out of connections for our account, I'd like us to log an entry in Sentry.

Comment (cadeParade) [reply app/services/websocket.js]: Have we confirmed that it works?

Comment (unicornzero) [reply app/services/websocket.js]: Discussed in chat. I ran into a snag and am pulling this out for now.

**Pull request merged**